pkg/controller/noderef/noderef_controller.go

```diff
@@ -0,0 +1,227 @@
+	if !machine.DeletionTimestamp.IsZero() {
+		klog.V(2).Infof("Machine %q in namespace %q has been deleted, won't reconcile", machine.Name, machine.Namespace)
+		return reconcile.Result{}, nil
+	}
+
+	// Check that the Machine doesn't already have a NodeRef.
+	if machine.Status.NodeRef != nil {
+		klog.V(2).Infof("Machine %q in namespace %q already has a NodeRef, won't reconcile", machine.Name, machine.Namespace)
+		return reconcile.Result{}, nil
+	}
+
+	// Check that the Machine has a cluster label.
+	if machine.Labels[v1alpha1.MachineClusterLabelName] == "" {
+		klog.V(2).Infof("Machine %q in namespace %q doesn't specify %q label, won't reconcile", machine.Name, machine.Namespace,
+			v1alpha1.MachineClusterLabelName)
+		return reconcile.Result{}, nil
+	}
+
+	// Check that the Machine has a valid Cluster associated with it.
+	cluster, err := r.getCluster(ctx, machine)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			klog.Warningf("Cannot find a Cluster for Machine %q in namespace %q, won't reconcile", machine.Name, machine.Namespace)
+			return reconcile.Result{}, err
+		}
+		return reconcile.Result{}, err
+	}
+
+	// Check that the Machine has a valid ProviderID.
+	if machine.Spec.ProviderID == nil || *machine.Spec.ProviderID == "" {
+		klog.Warningf("Machine %q in namespace %q doesn't have a valid ProviderID, retrying later", machine.Name, machine.Namespace)
+		return reconcile.Result{RequeueAfter: 30 * time.Second}, nil
+	}
+
+	result, err := r.reconcile(ctx, cluster, machine)
+	if err != nil {
+		klog.Errorf("Failed to assign NodeRef to Machine %q: %v", request.NamespacedName, err)
+		r.recorder.Event(machine, apicorev1.EventTypeWarning, "FailedSetNodeRef", err.Error())
+		return result, err
+	}
+
+	klog.Infof("Set Machine's (%q in namespace %q) NodeRef to %q", machine.Name, machine.Namespace, machine.Status.NodeRef.Name)
+	r.recorder.Event(machine, apicorev1.EventTypeNormal, "SuccessfulSetNodeRef", machine.Status.NodeRef.Name)
+	return result, nil
+}
+
+func (r *ReconcileNodeRef) reconcile(ctx context.Context, cluster *v1alpha1.Cluster, machine *v1alpha1.Machine) (reconcile.Result, error) {
+	providerID, err := noderefutil.NewProviderID(*machine.Spec.ProviderID)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	clusterClient, err := remote.NewClusterClient(r.Client, cluster)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	corev1Client, err := clusterClient.CoreV1()
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// Get the Node reference.
+	nodeRef, err := r.getNodeReference(corev1Client, providerID)
+	if err != nil {
+		if err == ErrNodeNotFound {
+			klog.Warningf("Cannot find a matching Node for Machine %q in namespace %q, retrying later", machine.Name, machine.Namespace)
+			return reconcile.Result{RequeueAfter: 10 * time.Second}, nil
+		}
+		return reconcile.Result{}, err
+	}
+
+	// Update Machine.
+	machine.Status.NodeRef = nodeRef
```

Comment:
There are currently workflows related to the existing Node and Machine controllers that will attempt to perform actions when the nodeRef is set.

- Node deletion: https://github.com/kubernetes-sigs/cluster-api/blob/bd0628f9aa808db1fdeed5c8bfba07ee2ecb1406/pkg/controller/machine/machine_controller.go#L182-L188
- isDeleteAllowed: https://github.com/kubernetes-sigs/cluster-api/blob/bd0628f9aa808db1fdeed5c8bfba07ee2ecb1406/pkg/controller/machine/machine_controller.go#L256-L275
- MachineSet status: https://github.com/kubernetes-sigs/cluster-api/blob/e325dfa6ecd8ed3ba68e2b07d645c8037c9fa09f/pkg/controller/machineset/status.go#L58-L63

Have these been tested with these changes?

Reply:
Those all fail as you’d expect. I’ll follow up with other PRs to tackle them separately